bayes_opt1/util.py

```diff
@@ -0,0 +1,547 @@
+            def func(w):
+                return mu_x_star_hat(x_obs,y_obs,gp,x_star_hat,mean,std,w)
+            
+            grad_temp.append(nd.Gradient(func)(x))
+    grad = np.zeros(p)
+    for i in grad_temp:
+        grad = grad + i
+    grad = grad/J
+    return grad
+
+
+
+# Implementation algorithm 3 (Frazier): Knowledge gradient 
+def minimize_kg_sgd(R, T, a, p, pbounds, optimizer, gp):
+
+    alpha = np.zeros(T)
+    KG = np.zeros(R)
+    temp = our_lhs(p,R,pbounds)
+    
+    if p == 1:
+        
+        x0 = np.zeros((T,R)) 
+        
+        for r in range(R):
+            # choose x0(r) uniformly random from A
+            t=0
+
+            x0[t,r] = temp[r]
+
+            for t in range(T-1):
+                #print(x0[(t+1)*r].reshape(1,-1))
+                G = _kg4(x0[t,r], optimizer, gp, n_grid = 50, J = 5)
+                alpha[t+1] = a/(a+t+1)
+                x0[t+1,r] = x0[t,r] + alpha[t+1]*G
+            #print(type(x0[T-1,r]))        
+            KG[r] = _kg2(x0[T-1,r], optimizer, gp, n_grid = 50, J = 5)
+
+
+    else:
+        x0 = []  
+
+        for r in range(R):
+            # choose x0(r) uniformly random from A
+            t=0
+
+            x0.append(temp[r].reshape(-1,1)) 
+
+            for t in range(T-1):
+                print(x0[(t+1)*r].reshape(1,-1))
+                G = _kg4(x0[(t+1)*r].reshape(1,-1), optimizer, gp, n_grid = 50, J = 5)
+                alpha[t+1] = a/(a+t+1)
+                x0.append(x0[(t+1)*r] + alpha[t+1]*G)
+            #print(type(x0[T-1,r]))        
+            KG[r] = _kg2(x0[(T-1)*r].reshape(1,-1), optimizer, gp, n_grid = 50, J = 5)
+
+    for i in range(R):
+        if KG[i] == np.max(KG):
+            return x0[i]
+    
+    
+
+def acq_max(optimizer, ac, gp, y_max, bounds, random_state, n_warmup=10000, n_iter=10, R=1, T=5, a=4):
+    """
+    A function to find the maximum of the acquisition function
+
+    It uses a combination of random sampling (cheap) and the 'L-BFGS-B'
+    optimization method. First by sampling `n_warmup` (1e5) points at random,
+    and then running L-BFGS-B from `n_iter` (250) random starting points.
+
+    Parameters
+    ----------
+    :param ac:
+        The acquisition function object that return its point-wise value.
+
+    :param gp:
+        A gaussian process fitted to the relevant data.
+
+    :param y_max:
+        The current maximum known value of the target function.
+
+    :param bounds:
+        The variables bounds to limit the search of the acq max.
+
+    :param random_state:
+        instance of np.RandomState random number generator
+
+    :param n_warmup:
+        number of times to randomly sample the aquisition function
+
+    :param n_iter:
+        number of times to run scipy.minimize
+
+    Returns
+    -------
+    :return: x_max, The arg max of the acquisition function.
+    """
+
+    p = bounds.shape[0]
+    
+     
+    if ac(None, optimizer, gp, y_max=0) == 'kg': 
```

Comment:
Abbiamo modificato la funzione Utility in modo che quando kind = kg utilizza effettivamente i tre algoritmi